fem/bilininteg_hcurl.cpp

```diff
@@ -366,6 +366,130 @@ void PAHcurlMassAssembleDiagonal3D(const int D1D,
    }); // end of element loop
 }
 
+void SmemPAHcurlMassAssembleDiagonal3D(const int D1D,
+                                       const int Q1D,
+                                       const int NE,
+                                       const Array<double> &_Bo,
+                                       const Array<double> &_Bc,
+                                       const Vector &_op,
+                                       Vector &_diag)
+{
+   constexpr static int VDIM = 3;
+   constexpr static int MAX_D1D = HCURL_MAX_D1D;
+   constexpr static int MAX_Q1D = HCURL_MAX_Q1D;
+
+   MFEM_VERIFY(D1D <= MAX_D1D, "Error: D1D > MAX_D1D");
+   MFEM_VERIFY(Q1D <= MAX_Q1D, "Error: Q1D > MAX_Q1D");
+
+   auto bo = Reshape(_Bo.Read(), Q1D, D1D-1);
+   auto bc = Reshape(_Bc.Read(), Q1D, D1D);
+   auto op = Reshape(_op.Read(), Q1D, Q1D, Q1D, 6, NE);
+   auto diag = Reshape(_diag.ReadWrite(), 3*(D1D-1)*D1D*D1D, NE);
+
+   MFEM_FORALL_3D(e, NE, Q1D, Q1D, Q1D,
+   {
+      MFEM_SHARED double BG[2][MAX_Q1D*MAX_D1D];
```

Comment:
All the shared arrays are of size `MAX_`*, is it because you dont instantiate any kernel?

Reply:
Yes, I am working on a template version with instantiation, which should be done soon (trying to debug some failing unit tests now).

Reply:
@camierjs I implemented a template version of this kernel in branch pacurlsmem-tmp, see commit 9cc330035f3f1d6fe584d494df441fa60c2b493e. This version passes the unit tests, but it is still using `MAX_D1D` although the use of the template argument `T_Q1D` is fine, see line 395 of bilininteg_hcurl.cpp in that commit. There is no difference in ex3p testing with this version, but a diagonal PA unit test fails for order 1 (D1D 2, Q1D 3). I cannot find the bug. Please let me know if you see something wrong. Also, I think this is the first time in MFEM that template functions are defined in a different cpp file, since the VectorFE kernels are split between H(curl) and H(div) files. The instantiations take more lines of code than I would like to have. If you see a better solution, I would be interested.

Reply:
@camierjs Update: I fixed the issue without understanding it, in commit 45457af5a625573576275aee7fbe84f46c21209d of pacurlsmem-tmp. It seems the declaration of `sBo` and `sBc` cannot be as offsets of `sBG`, although I don't know why. Declaring them separately as shared memory arrays works. Anyway, if you approve of this template version of the smem kernel, I will change the other smem kernels similarly in this PR branch.

Reply:
Ok - I think that the `tQ1D` should have been `tD1D` in `double (*sBo)[tQ1D] = (double (*)[tQ1D]) (sBG+0)`.
But your declaration is better.